### DISCUSSION TITLE
Building PetSc & 'Too many active git changes' & pyadjoint

### DISCUSSION THREAD

**Discussion by yianzeng** [firedrakeproject/firedrake#2605] [Firedrake support] [answered]
Hello! Running Python 3.10.7 (via homebrew) on an M1 Mac

I have 3 questions because I'm not sure where I've gone wrong, or if I've gone wrong everywhere.

1. Building PetSc
I get this message when it gets to the building PetSc part of the installation process
Depending on your platform, PETSc may take an hour or more to build!
Traceback (most recent call last):
  File "/Users/yianzeng/Desktop/coding/firedrake-install", line 1791, in <module>
    install("petsc/")
  File "/Users/yianzeng/Desktop/coding/firedrake-install", line 961, in install
    build_and_install_petsc()
  File "/Users/yianzeng/Desktop/coding/firedrake-install", line 1100, in build_and_install_petsc
    check_call([python, "./configure", "PETSC_DIR={}".format(petsc_dir), "PETSC_ARCH={}".format(petsc_arch)] + petsc_options)
  File "/Users/yianzeng/Desktop/coding/firedrake-install", line 592, in check_call
    log.debug(subprocess.check_output(arguments, stderr=subprocess.STDOUT, env=os.environ).decode())
  File "/opt/homebrew/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/opt/homebrew/Cellar/python@3.10/3.10.7/Frameworks/Python.framework/Versions/3.10/lib/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['/Users/yianzeng/Desktop/coding/firedrake/bin/python', './configure', 'PETSC_DIR=/Users/yianzeng/Desktop/coding/firedrake/src/petsc', 'PETSC_ARCH=default', '--download-hwloc-configure-arguments=--disable-opencl', '--with-blaslapack-dir=/opt/homebrew/opt/openblas', '--download-chaco', '--download-metis', '--with-debugging=0', '--with-x=0', '--download-hypre', '--download-bison', '--with-fortran-bindings=0', '--download-mumps', '--CFLAGS=-I/opt/homebrew/opt/openblas/include -Wno-implicit-function-declaration', '--download-suitesparse', '--with-cxx-dialect=C++11', '--download-mpich-configure-arguments=--disable-opencl', '--download-scalapack', '--download-mpich', '--download-netcdf', '--LDFLAGS=-Wl,-rpath,/opt/homebrew/opt/openblas/lib -L/opt/homebrew/opt/openblas/lib', '--download-pastix', '--download-ptscotch', '--with-zlib', '--with-c2html=0', '--with-shared-libraries=1', '--download-ml', '--download-hwloc', '--download-hdf5', '--download-superlu_dist', '--download-pnetcdf', '--download-eigen=/Users/yianzeng/Desktop/coding/firedrake/src/eigen-3.3.3.tgz ']' returned non-zero exit status 1.

I'm not quite sure what adjustments / arguments I should add in the installation stage.

2. I also tried to run the pytest as the repo has been partially cloned but VSC gives me the message that 'The git repository has too many active changes, only a subset of Git features will be enabled', and therefore I cannot run pytest.

3. I tried to gitclone the repository and run 'pytest tests/regression/ -k "poisson_strong or stokes_mini or dg_advection"' 
I got the following errors:

ImportError while loading conftest '/Users/yianzeng/Desktop/coding/firedrake/tests/conftest.py'.
tests/conftest.py:8: in <module>
    from pyadjoint.tape import get_working_tape
E   ModuleNotFoundError: No module named 'pyadjoint'

And 'pip3 install pyadjoint' led to:
ERROR: Could not find a version that satisfies the requirement pyadjoint (from versions: none)
ERROR: No matching distribution found for pyadjoint

[firedrake-install.log](https://github.com/firedrakeproject/firedrake/files/9868664/firedrake-install.log)
[configure.log](https://github.com/firedrakeproject/firedrake/files/9868739/configure.log)

I could not find a make.log file in my directory

**Comment by ReubenHill** [comment, marked as answer]
I am no expert on dealing with install issues, but I wouldn't be surprised if your issue is related to the warning you got early on in that log
```
Warning: Your Command Line Tools are too outdated.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.4.


Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.4.
```
I'll leave that to others to diagnose! Might be a good place to start though.

With regards to pytest not running, you need to have installed firedrake (which will install pyadjoint for you) before being able to run the firedrake tests.

**Reply by yianzeng** [reply]
Thank you for your answer! I shall try that now.

**Comment by ksagiyam** [comment]
I think that's right. We see `ld: library not found for -lSystem` in the log. We see this error periodically, and I think the solution was to reinstall command line tools.

**Reply by yianzeng** [reply]
Thank you!
